docs/src/using-turing/compiler.md

```diff
@@ -0,0 +1,340 @@
+---
+title: Turing Compiler Design
+---
+
+In this section, I will describe the current design of Turing's model compiler which enables Turing to perform various types of Bayesian inference without changing the model definition. What we call "compiler" is essentially just a macro that transforms the user's code to something that Julia's dispatch can operate on and that Julia's compiler can successfully do type inference on for efficient machine code generation.
+
+# Overview
+
+The following terminology will be used in this section:
+
+- `D`: observed data variables conditioned upon in the posterior,
+- `P`: parameter variables distributed according to the posterior distribution, these will also be referred to as random variables,
```

Comment:
Maybe better: distributed according to some prior distribution

Reply:
Sounds good.